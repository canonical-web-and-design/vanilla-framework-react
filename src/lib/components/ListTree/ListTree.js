import React from 'react';
import PropTypes from 'prop-types';
import './ListTree.scss';

const ListTree = (props) => {
  const listTreeChildren = React.Children.map(props.children, (child, index) =>
    React.cloneElement(child, { index: `${index}` }),
  );

  return (
    <ul className="p-list-tree" aria-multiselectable="true" role="tablist">
      {listTreeChildren}
    </ul>
  );
};

ListTree.propTypes = {
  children: PropTypes.node.isRequired,
};

ListTree.displayName = 'ListTree';

export default ListTree;
