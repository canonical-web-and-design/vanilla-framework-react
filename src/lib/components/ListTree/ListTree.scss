@import 'vanilla-framework/scss/base';
@import 'vanilla-framework/scss/patterns_list-tree';
@include vf-b-typography;
@include vf-p-list-tree;