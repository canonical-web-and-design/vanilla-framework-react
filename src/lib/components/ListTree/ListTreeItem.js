import React from 'react';
import PropTypes from 'prop-types';

const ListTreeItem = props => (
  <li key={props.index} className="p-list-tree__item">
    {props.children}
  </li>
);

ListTreeItem.defaultProps = {
  index: '0',
};

ListTreeItem.propTypes = {
  children: PropTypes.node.isRequired,
  index: PropTypes.string,
};

ListTreeItem.displayName = 'ListTreeItem';

export default ListTreeItem;
