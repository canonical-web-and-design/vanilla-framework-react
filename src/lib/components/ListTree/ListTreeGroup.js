import React from 'react';
import PropTypes from 'prop-types';
import './ListTree.scss';

class ListTreeGroup extends React.Component {
  constructor() {
    super();
    this.onClick = this.onClick.bind(this);
    this.state = {
      open: false,
    };
  }

  onClick() {
    const { open } = this.state;
    this.setState({ open: !open });
  }

  render() {
    const { index } = this.props;
    const listTreeGroupChildren = React.Children.map(this.props.children, (child, key) =>
      React.cloneElement(child, { index: `${index}-${key}` }),
    );

    return (
      <li className="p-list-tree__item p-list-tree__item--group">
        <button
          className="p-list-tree__toggle"
          id={`listtree-${index}-btn`}
          role="tab"
          aria-controls={`listtree-${index}`}
          aria-expanded={this.state.open}
          onClick={this.onClick}
        >
          {this.props.name}
        </button>
        <ul
          className="p-list-tree"
          id={`listtree-${index}`}
          role="tabpanel"
          aria-hidden={!(this.state.open)}
          aria-labelledby={`listtree-${index}-btn`}
        >
          {listTreeGroupChildren}
        </ul>
      </li>
    );
  }
}

ListTreeGroup.defaultProps = {
  index: '0',
};

ListTreeGroup.propTypes = {
  children: PropTypes.node.isRequired,
  index: PropTypes.string,
  name: PropTypes.string.isRequired,
};

ListTreeGroup.displayName = 'ListTreeGroup';

export default ListTreeGroup;
