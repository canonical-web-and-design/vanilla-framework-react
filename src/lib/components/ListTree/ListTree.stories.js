import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import ListTree from './ListTree';
import ListTreeGroup from './ListTreeGroup';
import ListTreeItem from './ListTreeItem';
import Link from '../Link/Link';
import Image from '../Image/Image';

storiesOf('List Tree', module)
  .add('Default',
    withInfo('The ListTree component can be used to show a directory style listing, such as a list of files and folders within a directory.')(() => (
      <ListTree>
        <ListTreeGroup name="/folder">
          <ListTreeItem>file</ListTreeItem>
        </ListTreeGroup>
        <ListTreeItem>
          <Link strong href="#a">charm-helpers-sync.yaml</Link>
        </ListTreeItem>
        <ListTreeItem>
          <Link strong href="#a">config.yaml</Link>
        </ListTreeItem>
        <ListTreeGroup name="/files">
          <ListTreeItem>default_rsync</ListTreeItem>
          <ListTreeItem>nagios_plugin.py</ListTreeItem>
          <ListTreeGroup name="/plugins">
            <ListTreeItem>check_mem.pl</ListTreeItem>
          </ListTreeGroup>
          <ListTreeGroup name="/misc">
            <ListTreeGroup name="/cat">
              <ListTreeItem>
                <Image bordered src="http://placekitten.com/g/300/300" alt="" />
              </ListTreeItem>
            </ListTreeGroup>
          </ListTreeGroup>
        </ListTreeGroup>
      </ListTree>),
    ),
  );
