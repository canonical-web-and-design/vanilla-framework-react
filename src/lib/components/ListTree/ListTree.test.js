import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import ListTree from './ListTree';
import ListTreeGroup from './ListTreeGroup';
import ListTreeItem from './ListTreeItem';

describe('ListTree component', () => {
  it('should render a basic ListTree correctly', () => {
    const listTree = ReactTestRenderer.create(
      <ListTree>
        <ListTreeGroup name="/group">
          <ListTreeItem>file</ListTreeItem>
        </ListTreeGroup>
      </ListTree>);
    const json = listTree.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a nested ListTree correctly', () => {
    const listTree = ReactTestRenderer.create(
      <ListTree>
        <ListTreeGroup name="/depth1">
          <ListTreeGroup name="/depth2">
            <ListTreeGroup name="/depth3">
              <ListTreeItem>file</ListTreeItem>
            </ListTreeGroup>
          </ListTreeGroup>
        </ListTreeGroup>
      </ListTree>);
    const json = listTree.toJSON();
    expect(json).toMatchSnapshot();
  });
});
