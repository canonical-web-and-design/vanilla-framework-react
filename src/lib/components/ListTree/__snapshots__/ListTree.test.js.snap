// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListTree component should render a basic ListTree correctly 1`] = `
<ul
  aria-multiselectable="true"
  className="p-list-tree"
  role="tablist"
>
  <li
    className="p-list-tree__item p-list-tree__item--group"
  >
    <button
      aria-controls="listtree-0"
      aria-expanded={false}
      className="p-list-tree__toggle"
      id="listtree-0-btn"
      onClick={[Function]}
      role="tab"
    >
      /group
    </button>
    <ul
      aria-hidden={true}
      aria-labelledby="listtree-0-btn"
      className="p-list-tree"
      id="listtree-0"
      role="tabpanel"
    >
      <li
        className="p-list-tree__item"
      >
        file
      </li>
    </ul>
  </li>
</ul>
`;

exports[`ListTree component should render a nested ListTree correctly 1`] = `
<ul
  aria-multiselectable="true"
  className="p-list-tree"
  role="tablist"
>
  <li
    className="p-list-tree__item p-list-tree__item--group"
  >
    <button
      aria-controls="listtree-0"
      aria-expanded={false}
      className="p-list-tree__toggle"
      id="listtree-0-btn"
      onClick={[Function]}
      role="tab"
    >
      /depth1
    </button>
    <ul
      aria-hidden={true}
      aria-labelledby="listtree-0-btn"
      className="p-list-tree"
      id="listtree-0"
      role="tabpanel"
    >
      <li
        className="p-list-tree__item p-list-tree__item--group"
      >
        <button
          aria-controls="listtree-0-0"
          aria-expanded={false}
          className="p-list-tree__toggle"
          id="listtree-0-0-btn"
          onClick={[Function]}
          role="tab"
        >
          /depth2
        </button>
        <ul
          aria-hidden={true}
          aria-labelledby="listtree-0-0-btn"
          className="p-list-tree"
          id="listtree-0-0"
          role="tabpanel"
        >
          <li
            className="p-list-tree__item p-list-tree__item--group"
          >
            <button
              aria-controls="listtree-0-0-0"
              aria-expanded={false}
              className="p-list-tree__toggle"
              id="listtree-0-0-0-btn"
              onClick={[Function]}
              role="tab"
            >
              /depth3
            </button>
            <ul
              aria-hidden={true}
              aria-labelledby="listtree-0-0-0-btn"
              className="p-list-tree"
              id="listtree-0-0-0"
              role="tabpanel"
            >
              <li
                className="p-list-tree__item"
              >
                file
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
`;
